profiling/src/profiling/mod.rs

```diff
@@ -510,7 +516,40 @@ pub enum UploadMessage {
 #[cfg(feature = "timeline")]
 const COW_EVAL: Cow<str> = Cow::Borrowed("[eval]");
 
+const DDPROF_TIME: &str = "ddprof_time";
+const DDPROF_UPLOAD: &str = "ddprof_upload";
+
+static PROFILER_INIT: Mutex<()> = Mutex::new(());
+
 impl Profiler {
+    /// Will initialize the `PROFILER` OnceCell and makes sure that only one thread will do so.
+    pub fn init(system_settings: &mut SystemSettings) {
+        // Safety: there could be a race in which another thread already initializes `PROFILER`,
+        // this is guarded by the `PROFILER_INIT` mutex next.
+        if unsafe { PROFILER.get() }.is_some() {
+            return;
+        }
```

Comment:
This is definitely race condition. Whether it's one of those harmless ones is a bit trickier for me to think about because I don't know Rust's internals of these things as well as I do C's, where I spent most of my time learning correct parallelism. I assume you wish to avoid touching the lock if it's already initialized for performance reasons?

Reply:
> I assume you wish to avoid touching the lock if it's already initialized for performance reasons?

Exactly

Reply:
I think you don't need to worry about it, on Linux the `futex` uses an atomic relaxed load to determine if the lock is contended: https://github.com/rust-lang/rust/pull/95035/files#diff-823e207a741b9946be4e20738fd2cfe16e077beb001632aaa8623e3ea0803cc0. I think you can just delete these lines.

Reply:
I guess there's still the mutex poisoning wrapper around it. I wonder if we should use `parking_lot`'s `Mutex` instead.

Reply:
I need to be more precise about this code maybe:
- this is called from `RINIT` just to make it explicit (we could also check in the `Profiler::get()` method if the `Profiler` is initialised and initialise it in case it is not, but this way is more explicit)
- in case the Profiler is already initialised (`PROFILER` is some), I do not need to to anything anymore
- in case it is not, I need to initialise it, this is where the `PROFILER_INIT` mutex comes into play, because I only want one thread to actually initialise the `Profiler` and all others to wait